### PR TITLE
adding support for talos.dev clusters iscsi

### PR DESCRIPTION
**Pull request by 5cat**
Hi, I have added some parameters to allow for [talos.dev](https://www.talos.dev/) iscsi support.

Appreciate your feedback

### CODE REVIEW

**Comment by 5cat** [comment]
This pull request is related to https://github.com/democratic-csi/charts/pull/34

**Comment by travisghansen** [comment]
This appears to have taken a bit of research and trial/error to get going, nice work! I don't have any experience with `talos` but the instructions seem sane enough so I'll merge after the chart PR merges.

**Comment by 5cat** [comment]
Thank you @travisghansen

**Comment by travisghansen** [comment]
Let's add a `ISCSIADM_HOST_STRATEGY=chroot|nsenter` env var to the wrapper and build out the logic as appropriate. That should simplify things.

**Comment by travisghansen** [comment]
Cross reference here as well: https://github.com/siderolabs/extensions/issues/38

**Comment by 5cat** [comment]
Hey @travisghansen , 

I built the image and I'm getting these errors when I test it with a pvc for a new pod. Any idea as for why?
```
  Normal   Scheduled    2m48s  default-scheduler  Successfully assigned test-democratic-csi/terminal-75cf46db49-ln65s to talos-192-168-0-125
  Warning  FailedMount  2m47s  kubelet            MountVolume.MountDevice failed for volume "pvc-7c35b2ef-65ce-4fa3-af29-399bbc003b71" : rpc error: code = Internal desc = TypeError: Cannot read properties of undefined (reading 'split') TypeError: Cannot read properties of undefined (reading 'split')
    at /home/csi/app/src/utils/iscsi.js:240:31
    at Array.forEach (<anonymous>)
    at Object.getSessions (/home/csi/app/src/utils/iscsi.js:232:17)
    at processTicksAndRejections (node:internal/process/task_queues:96:5)
    at async Object.getSession (/home/csi/app/src/utils/iscsi.js:179:26)
    at async FreeNASApiDriver.NodeStageVolume (/home/csi/app/src/driver/index.js:779:29)
    at async requestHandlerProxy (/home/csi/app/bin/democratic-csi:217:18)
  Warning  FailedMount  46s                  kubelet  Unable to attach or mount volumes: unmounted volumes=[test-stuff], unattached volumes=[test-stuff kube-api-access-khhx2]: timed out waiting for the condition
  Warning  FailedMount  27s (x8 over 2m45s)  kubelet  MountVolume.MountDevice failed for volume "pvc-7c35b2ef-65ce-4fa3-af29-399bbc003b71" : rpc error: code = Internal desc = TypeError: Cannot read properties of undefined (reading 'split') TypeError: Cannot read properties of undefined (reading 'split')
    at /home/csi/app/src/utils/iscsi.js:240:31
    at Array.forEach (<anonymous>)
    at Object.getSessions (/home/csi/app/src/utils/iscsi.js:232:17)
    at runMicrotasks (<anonymous>)
    at processTicksAndRejections (node:internal/process/task_queues:96:5)
    at async Object.getSession (/home/csi/app/src/utils/iscsi.js:179:26)
    at async FreeNASApiDriver.NodeStageVolume (/home/csi/app/src/driver/index.js:779:29)
    at async requestHandlerProxy (/home/csi/app/bin/democratic-csi:217:18)
    ```
it is erroring on this line  `response = await driver[serviceMethodName](call);`
My other previsioner is running and i'm using freenas-api-iscsi driver.

File: README.md [hunk 1]
```diff
@@ -134,6 +135,45 @@ sudo systemctl enable open-iscsi.service
...
+```yaml
+
+node:
+  hostPID: true
+  driver:
+    extraEnv:
+      - name: ISCSIADM_HOST_STRATEGY
+        value: nsenter
+      - name: ISCSIADM_HOST_PATH
+        value: /usr/local/sbin/iscsiadm
```
Comment (5cat) [inline README.md:171]: @travisghansen  Do you prefer if I made these envs in the `node.yaml` file in the charts repo with a key `node.driver.iscsiadmHostStrategy` and `node.driver.iscsiadmHostPath` ?

Comment (travisghansen) [reply README.md:171]: No just common env vars is preferred for now.

**Comment by travisghansen** [comment]
That failure is likely because of the echo statements you added to the wrapper. There are many places where I parse the output of the iscsiadm command so those will need to be removed.

**Comment by 5cat** [comment]
@travisghansen I have tested both repos now with my freenas iscsi driver and it seems working as expected. May I ask if you can test the chart on your side with old iscsi config just to make sure my changes are backward compatible?

**Comment by travisghansen** [comment]
I'll merge this into `next` where I'll tweak the syntax a bit. Is there a reason you didn't use the `pgrep` tool for getting the pid?

**Comment by 5cat** [comment]
Yes please feel free to improve it. the `pgrep` tool wasn't available, nether does `ps` or other tools I know of, so the only thing left for me is to do it the raw way.  But I guess we can install `pgrep` in the `Dockerfile` to make it more clean.

**Comment by travisghansen** [comment]
Yeah, I'll throw `pgrep` into the image.

**Comment by travisghansen** [comment]
I've set these as the defaults (images are building currently):

```
: "${ISCSIADM_HOST_STRATEGY:=chroot}"
: "${ISCSIADM_HOST_PATH:=iscsiadm}"
```

So for your purposes you'll need to override both if you need to give the full path to `nsenter` (which I don't actually think you need to do).

As an example from my local workstation:

```
sudo ISCSIADM_HOST_STRATEGY=nsenter ./docker/iscsiadm -m session
iscsiadm: No active sessions.
```

**Comment by travisghansen** [comment]
OK, new images built with the `next` tag. Give it a try and let's see how 'official' images do.

**Comment by 5cat** [comment]
Hi @travisghansen , I can confirm the freenas-api-iscsi driver works on talos with this image `docker.io/democraticcsi/democratic-csi:next` and this helm [chart branch](https://github.com/democratic-csi/charts/pull/34).

My method of testing it is simply creating an alpine container with a pvc and see if the pvc is created and im able to create files in it.

**Comment by travisghansen** [comment]
Ok I’ll merge to master and snap a release as well.

**Comment by 5cat** [comment]
Thank you @travisghansen for the help, looking forward to use the official release.

**Comment by travisghansen** [comment]
`v1.7.6` released 

Now I’ll look at the chart more closely.

**Comment by Twanislas** [comment]
Hi folks !

Reviving this "thread" as it looks like multipath support would need the same "nsenter" treatment to be able to work on Talos.

I submitted an issue to the Talos team to add multipath to their iscsi-tools extentions as it looks like it's missing... See https://github.com/siderolabs/extensions/issues/134

I will try to have a stab at it but I'm not sure I'd have the skills to make a nice PR :)

Thanks for taking this into consideration !

**Comment by travisghansen** [comment]
multipath 'just works' behind the scenes, I don't execute any multipath commands or anything of that nature so nsenter isn't really relevant (multipath effectively operates entirely in the kernel).

**Pull request merged**